mclient/mkds.8
.\"
.\"
.\"
.\"
.\"
.\"
.TH MKDS 8 "16 October 2009" "MIT SIPB"
.SH NAME
mkds \- Create a discuss meeting
.SH SYNOPSIS
.B mkds
.SH DESCRIPTION
.I mkds
creates a discuss meeting on a specified discuss server,
provided the user has permission to do so.

.I mkds 
will prompt for the server name, a directory where the meeting should
be stored, the long and short names of the meeting, and whether or not
the meeting should be public.
At this point, it attempts to contact the discuss daemon running on the
discuss server, and an access checked is performed for meeting creation
(the ``cacl'' file on the server lists authorized users).
If the meeting can be successfully created and is not a public meeting,
.I mkds
will then prompt for a list of users who should have
permission to create new transactions in the meeting.
(The user who created the meeting will automatically be given
chairman permissions.)
.I mkds
will then open an editor for
the creator to enter the initial meeting transaction.  Finally,
.I mkds
will ask whether or not the meeting should be announced
(in the new_meetings meeting).

.I mkds
accepts no arguments.

The meeting created by
.I mkds
is a pure discuss meeting in its original form.  If it is intended
to be used as a mail archive, a server maintainer will need to perform
additional configuration to accept mail for the meeting's alias and
run
.I dsmail
as appropriate.

A meeting created with
.I mkds
can be removed with
\fIrmds\fR.

The cacl file has a similar format to that of a discuss meeting acl;
the first line contains a number which is the number of entries on the
acl.
The subsequent lines contain a list of access bits, a colon,
and the user they correspond to.
However, unlike meeting acls, only the 'a' bit is relevant for
meeting creation.

.SH FILES
.ta \w'/var/spool/discuss/cacl  'u
.br
~/.meetings	Data file listing meetings that you attend.
.br
/var/spool/discuss/cacl	The list of users (Kerberos principals) who 
	can create meetings on the server.

.SH AUTHORS
Robert French and Ken Raeburn, MIT Student Information Processing Board

.SH "SEE ALSO"
discuss(1), dsmail(1), rmds(8)








debian/changelog
debathena-discuss (10.0.13-0debathena2) unstable; urgency=low

  * Flesh out mkds.8 and rmds.8 some more.

 -- Benjamin Kaduk <kaduk@mit.edu>  Sun, 03 Jul 2011 18:31:43 -0400

debathena-discuss (10.0.13-0debathena1) unstable; urgency=low

  * Update our autogoo for the 21st century.
  * Add a working distclean target.
  * Update packaging to not require daconfiscate.
  * Eliminate libds, as its only use could be replaced by libdsk.
  * Check $PAGER for a pager before falling back on a compiled-in default
    ("more" by default, but "pager" under our packaging). (Trac: #629)

 -- Evan Broder <broder@mit.edu>  Sun, 27 Jun 2010 13:02:34 -0400

debathena-discuss (10.0.12-0debathena3) unstable; urgency=low

  * Don't depend on tcsh anymore

 -- Jonathan Reed <jdreed@mit.edu>  Thu, 08 Apr 2010 22:41:16 -0400

debathena-discuss (10.0.12-0debathena2) unstable; urgency=low

  * Run dsc_setup on behalf of the user, like edsc does. (Trac #426)
  * Rewrite dsc_setup in bash, because all the cool kids use it.

 -- Jonathan Reed <jdreed@mit.edu>  Thu, 08 Apr 2010 21:03:07 -0400

debathena-discuss (10.0.12-0debathena1) unstable; urgency=low

  * Stop using the obsolete AC_TYPE_SIGNAL. (Trac: #347)
  * Add missing manpages. (Trac: #266)

 -- Evan Broder <broder@mit.edu>  Sun, 15 Nov 2009 17:35:19 -0500

debathena-discuss (10.0.11-0debathena1) unstable; urgency=low

  * Move the single-letter checks in trn_expr.y:yylex to the bottom of the
    function so that they get caught after commands that start with those
    letters. (Trac: #346)

 -- Evan Broder <broder@mit.edu>  Sat, 29 Aug 2009 17:20:43 -0400

debathena-discuss (10.0.10-0debathena2) unstable; urgency=low

  * Depend on Kerberos-enabled zephyr instead of build-depending.

 -- Evan Broder <broder@mit.edu>  Mon, 27 Jul 2009 14:38:10 -0400

debathena-discuss (10.0.10-0debathena1) unstable; urgency=low

  * Patch discuss server to work without krb4. Patch by Mitch Berger.

 -- Evan Broder <broder@mit.edu>  Thu, 23 Jul 2009 21:00:55 -0400

debathena-discuss (10.0.9-0debathena1) unstable; urgency=low

  * Fix local/dumb authentication needed for mail delivery and local
    client usage. Patch by Mitch Berger. (Trac: #274)
  * Always build without krb4.

 -- Evan Broder <broder@mit.edu>  Wed, 22 Jul 2009 22:58:40 -0400

debathena-discuss (10.0.8-0debathena1) unstable; urgency=low

  * Add one-character aliases to discuss. Patch by Ben Kaduk. (Trac: #238)
  * Squash some lintian warnings by switching from ${Source-Version} to
    ${source:Version} and ${binary:Version}
  * Use HAVE_KRB4 and HAVE_KRB5 instead of KERBEROS and KERBEROS5.
  * Assorted changes to make discuss work with any combination of krb4 and
    krb5 (including neither).
  * Always build our discuss with krb5.

 -- Evan Broder <broder@mit.edu>  Mon, 18 May 2009 23:51:52 -0400

debathena-discuss (10.0.7-0debathena3) unstable; urgency=low

  * Configure the discuss inetd service.

 -- Evan Broder <broder@mit.edu>  Thu, 05 Mar 2009 15:17:20 -0500

debathena-discuss (10.0.7-0debathena2) unstable; urgency=low

  * Change DEB_AUTO_UPDATE_AUTOCONF to 2.50, not 1.
  * Remove DEB_AUTO_UPDATE_DEBIAN_CONTROL.
  * Change "Debian-Athena Project" to "Debathena Project".
  * Remove debian/control from version control.
  * Clean up debian/copyright.

 -- Tim Abbott <tabbott@mit.edu>  Sat, 24 Jan 2009 14:31:41 -0500

debathena-discuss (10.0.7-0debathena1) unstable; urgency=low

  * Revert a small part of the OS X com_err patch from 10.0.6 in favor of
    changes to aclocal.m4
  * Remove hard-coded paths from discuss-mode, in favor of getting them
    from autoconf.
  * Remove debathena-machtype as a dependency, as it's no longer needed.

 -- Evan Broder <broder@mit.edu>  Fri, 09 Jan 2009 21:47:33 -0500

debathena-discuss (10.0.6-0debathena1) unstable; urgency=low

  [ Evan Broder ]
  * Add com_err support for delete and discuss on OS X.

  [ Anders Kaseorg ]
  * Depend emacs | emacsen instead of emacs22 | emacs21 | emacsen, so that
    the emacs metapackage is pulled in by default.

 -- Anders Kaseorg <andersk@mit.edu>  Thu, 08 Jan 2009 16:35:04 -0500

debathena-discuss (10.0.5-0debathena1) unstable; urgency=low

  * Fix some trivial typos.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 19 Aug 2008 16:35:07 -0400

debathena-discuss (10.0.4-0debathena1) unstable; urgency=low

  * Fix a crash bug in edsc on 64-bit platforms.
  * Fix date display in discuss on 64-bit platforms.

 -- Greg Hudson <ghudson@mit.edu>  Mon, 30 Jun 2008 12:50:56 -0400

debathena-discuss (10.0.3-0debathena1) unstable; urgency=low

  * Fix a crash bug in edsc (int/long mismatch in short_time).
  * Put edsc crash dumps in /tmp instead of /usr/tmp.

 -- Greg Hudson <ghudson@mit.edu>  Wed, 11 Jun 2008 19:39:38 -0400

debathena-discuss (10.0.2-0debathena1) unstable; urgency=low

  * Stop assuming that long is 32 bits.
  * Start assuming that int is 32 bits.

 -- Greg Hudson <ghudson@mit.edu>  Wed, 11 Jun 2008 16:29:35 -0400

debathena-discuss (10.0.1-0debathena1) unstable; urgency=low

  * Fix obsolete names for Kerberos library functions, which no longer
    worked on Mac OS X.  Patch from broder.

 -- Anders Kaseorg <andersk@mit.edu>  Sun, 06 Apr 2008 19:25:13 -0400

debathena-discuss (10.0.0-0debathena2) unstable; urgency=low

  * Actually use SERVERDIR, BINDIR, and INFODIR in config.h.sed.  (The
    Makefile already interpolates these variables.)

 -- Anders Kaseorg <andersk@mit.edu>  Sun, 06 Apr 2008 17:28:52 -0400

debathena-discuss (10.0.0-0debathena1) unstable; urgency=low

  * Merged quilt patches into mainline Athena tree

 -- Shreevatsa R <vatsa@mit.edu>  Sun,  6 Apr 2008 15:54:22 -0400

debathena-discuss (9.4.41-0debathena2) unstable; urgency=low

  * Add emacs22 as a default alternative to emacs21 | emacsen.

 -- Anders Kaseorg <andersk@mit.edu>  Mon, 29 Oct 2007 21:40:29 -0400

debathena-discuss (9.4.41-0debathena1) unstable; urgency=low

  * New upstream release supporting Kerberos 5.


mclient/mkds.c
/*
 *
 *	Copyright (C) 1988, 1989 by the Massachusetts Institute of Technology
 *    	Developed by the MIT Student Information Processing Board (SIPB).
 *    	For copying information, see the file mit-copyright.h in this release.
 *
 */
/*
 *	$Id: mkds.c,v 1.23 1999-02-08 14:47:15 danw Exp $
 *
 */

#ifndef lint
static char rcsid_mkds_c[] =
    "$Id: mkds.c,v 1.23 1999-02-08 14:47:15 danw Exp $";
#endif /* lint */

#include <discuss/discuss.h>
#if 0
#include "dsc_et.h"
#include "config.h"
#include "interface.h"
#include "rpc.h"
#include "globals.h"
#endif
#include <sys/time.h>
#include <sys/file.h>
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <netdb.h>
#include <pwd.h>
#include <errno.h>
#if HAVE_FCNTL_H
#include <fcntl.h>
#endif

#define cupper(x) (islower(x)?toupper(x):(x))
#define clower(x) (isupper(x)?tolower(x):(x))

char default_dir[] = "/var/spool/discuss";
char *whoami;
int  interrupt = 0;

extern const char *local_realm();

char *strtrim();

main(argc,argv)
int argc;
char *argv[];
{
	extern tfile unix_tfile();
	name_blk nbsrc,nbdest;
	char long_name[100],short_name[100],username[50],mtg_path[100];
	char mtg_host[100];
	char *default_host;
	char temp_file[64];
	char ann_mtg[100];
	int public = 0,error = 1,result,remove=0,delmtg=0;
	int fd,txn_no;
	tfile tf;
	char hostname[256];

#if defined(__APPLE__) && defined(__MACH__)
	add_error_table(&et_dsc_error_table);
#else
	initialize_dsc_error_table();
#endif

	nbsrc.user_id = malloc(132);

	(void) sprintf(temp_file,"/tmp/mtg%d.%d",getuid(),getpid());

	whoami = strrchr(argv[0],'/');
	if (whoami)
		whoami++;
	else
		whoami = argv[0];

	if (argc > 1) {
		fprintf(stderr,"Usage: %s\n",whoami);
		exit (1);
	}

	if (!strcmp(whoami,"rmds"))
		remove++;
	else if (strcmp(whoami, "mkds")) {
		fprintf(stderr,
			"This program must be run as 'mkds' or 'rmds'.\n");
		exit(1);
	}

	gethostname(hostname, 256);
	{
		struct hostent *hp;
		char *h;
		hp = gethostbyname(hostname);
		h = (hp ? hp->h_name : hostname);
		default_host = malloc(strlen(h)+1);
		strcpy(default_host, h);
	}
	printf("Meeting host: [default %s]: ", default_host);
	if (!gets(mtg_host))
		exit(1);
	strcpy(mtg_host, strtrim(mtg_host));
	if (mtg_host[0] == '\0')
		strcpy(mtg_host, default_host);
	if (mtg_host[0] == '%')
		strcpy(mtg_host, "");
	else {
		struct hostent *hp;
		hp = gethostbyname(mtg_host);
		if (!hp) {
			fprintf(stderr, "Unknown host %s\n", mtg_host);
			exit(1);
		}
		strcpy(mtg_host, hp->h_name);
	}
	printf("\nMeeting location [default %s]: ", default_dir);
	if (!gets(mtg_path))
	        exit(1);
	strcpy(mtg_path, strtrim(mtg_path));
	if (!mtg_path[0])
		strcpy(mtg_path, default_dir);
	if (!remove) {
		printf("\nLong meeting name: ");
		if (!gets(long_name))
		        exit(1);
		strcpy(long_name, strtrim(long_name));
		if (long_name[0] == '\0') {
			printf("No long meeting name supplied.\n");
			exit(1);
		}
	}
	printf("\nShort meeting name: ");
	if (!gets(short_name))
	        exit(1);
	strcpy(short_name, strtrim(short_name));
	if(short_name[0] == '\0') {
		printf("No short meeting name supplied.\n");
		exit(1);
	}
	
	(void) strcpy(mtg_path, strtrim(mtg_path));
	(void) strcat(mtg_path,"/");
	(void) strcat(mtg_path, short_name);
	nbsrc.pathname = malloc(strlen(mtg_path) + 1);
	strcpy(nbsrc.pathname, mtg_path);

	nbsrc.hostname = malloc(strlen(mtg_host) + 1);
	strcpy(nbsrc.hostname, mtg_host);

	(void) strcpy (username, getpwuid(getuid())->pw_name);
	(void) strcpy (nbsrc.user_id, username);

	if (remove) {
		dsc_remove_mtg(&nbsrc,&result);
		if (result)
			(void) fprintf(stderr,"Can't remove meeting: %s\n",
				       error_message(result));
		error = result;
		goto kaboom;
	}

	printf("\n");
	public = getyn("Should this meeting be public [y]? ",'Y');

	dsc_create_mtg(mtg_host, mtg_path, long_name, public, 0,
		       &result);
	if (result) {
		if (result == ECONNREFUSED)
			fprintf(stderr, "%s doesn't appear to be running a discuss server", mtg_host);
		else fprintf(stderr, "%s.  Can't create meeting.\n",
			     error_message(result));
		goto kaboom;
	}

	if (!public &&
	    getyn("Should specified users be allowed to participate? [y]", 'Y')) {
		char username[140];
		
		printf("Enter the usernames you wish to participate; \n\
End with . on a line by itself\n\n");
		for (;;) {
			printf("User name: ");
			fflush(stdout);
			if (!gets(username)) break;
			strcpy(username, strtrim(username));
			if (strcmp(username, ".") == 0) break;
			if (strcmp(username,"*") != 0 &&
			    strchr(username, '@') == 0) {
				strcat(username, "@");
				strcat(username, local_realm());
			}
			dsc_set_access (&nbsrc, username, "aorsw", &result);
			if (result) {
				fprintf (stderr, "Can't add participant: %s\n",
					 error_message(result));
			}
		}
	}
	clearerr(stdin);
	
	delmtg = 1;

	nbsrc.date_attended = time(0);
	nbsrc.last = 0;
	nbsrc.status = 0;
	nbsrc.aliases = (char **) calloc(3, sizeof(char *));
	nbsrc.aliases[0] = malloc(strlen(long_name)+1);
	strcpy(nbsrc.aliases[0], long_name);
	nbsrc.aliases[1] = malloc(strlen(short_name)+1);
	strcpy(nbsrc.aliases[1], short_name);
	nbsrc.aliases[2] = (char *)NULL;

	nbsrc.spare = "";
		
	dsc_update_mtg_set(username,&nbsrc,1,&result);
	if (result) {
		fprintf(stderr, "mkds: Can't set meeting name: %s",
			error_message(result));
		goto kaboom;
	}

	printf("\nYou must now enter the initial transaction.\n");
	printf(
	  "This transaction will serve as an introduction to the meeting.\n");

	(void) unlink(temp_file);

	if (edit(temp_file, getenv("EDITOR"))) {
		(void) fprintf(stderr,
		  "Error during edit; transaction not entered.\n");
		goto kaboom;
	}

	fd = open(temp_file,O_RDONLY,0);
	if (fd < 0) {
		(void) fprintf(stderr,"No file; not entered.\n");
		goto kaboom;
	}
	tf = unix_tfile(fd);

	dsc_add_trn(&nbsrc, tf, "Reason for this meeting", 0, &txn_no,
		    &result);
	if (result) {
		fprintf(stderr, "mkds: Error adding transaction: %s",
			error_message(result));
		goto kaboom;
	}
	(void) printf("Transaction [%04d] entered in the %s meeting.\n",
		txn_no,long_name);

	(void) close(fd);

	printf("\n");
	if (getyn("Would you like to announce this meeting [y]? ",'Y')) {
		printf("\n");
		for (;;) {
			printf("Announce in what meeting? ");
			if (!gets(ann_mtg))
				exit(1);
			dsc_get_mtg(nbsrc.user_id,strtrim(ann_mtg),
				    &nbdest,&result);
			if (!result)
				break;
			fprintf(stderr, "Meeting not found in search path.\n");
		}

		fd = open(temp_file,O_RDONLY,0);
		if (fd < 0) {
			(void) fprintf(stderr,"Temporary file disappeared!\n");
			goto kaboom;
		}

		tf = unix_tfile(fd);

		dsc_announce_mtg(&nbsrc, &nbdest, public, tf,
				 &txn_no, &result);

		if (result) {
			(void) fprintf(stderr,
				       "mkds: Error adding transaction: %s\n",
				       error_message(result));
			(void) fprintf(stderr,
			   "Use the announce_meeting (anm) request in discuss.\n");
		}
		else (void) printf("Transaction [%04d] entered in the %s meeting.\n",
				   txn_no, nbdest.aliases[0]);

		(void) close(fd);
	}
		
	error = 0;

kaboom:

	(void) unlink(temp_file);

	if (error && delmtg) {
		fprintf(stderr,"\nError encountered - deleting meeting.\n");
		remove_mtg(mtg_path,&result);
		if (result)
		     perror("Can't delete meeting");
	}
	term_rpc();
	exit(!error);
}

getyn(prompt,def)
char *prompt,def;
{
	char yn_inp[128];

	for (;;) {
		(void) printf("%s ",prompt);
		if (!gets(yn_inp))
		        exit(1);
		if (yn_inp[0] == '\0')
			yn_inp[0] = def;
		if (cupper(yn_inp[0]) == 'Y' || cupper(yn_inp[0]) == 'N')
			return (cupper(yn_inp[0]) == 'Y');
		printf("Please enter 'Yes' or 'No'\n\n");
	}
}
#include <ctype.h>

char *strtrim(cp)
	register char *cp;
{
	register int c;
	register char *cp1;

	while ((c = *cp) && isspace (c)) cp++;
	cp1 = cp;
	while (*cp1) cp1++;
	do {
		cp1--;
	} while (cp1 > cp && isspace (*cp1));
	cp1++;
	*cp1 = '\0';
	return cp;
}
